receiver/hostmetricsreceiver/internal/scraper.go

```diff
@@ -16,24 +16,26 @@ package internal
 
 import (
 	"context"
-	"time"
 
 	"go.uber.org/zap"
 
-	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/pdata"
 )
 
 // Scraper gathers metrics from the host machine and converts
 // these into internal metrics format.
 type Scraper interface {
-	// Start performs any timely initialization tasks such as
-	// setting up performance counters for initial collection,
-	// and then begins scraping metrics at the configured
-	// collection interval.
-	Start(ctx context.Context) error
-	// Shutdown should clean up any unmanaged resources such as
+	// Initialize performs any timely initialization tasks such as
+	// setting up performance counters for initial collection.
+	Initialize(ctx context.Context) error
+	// Close should clean up any unmanaged resources such as
 	// performance counter handles.
-	Shutdown(ctx context.Context) error
+	Close(ctx context.Context) error
+	// ScrapeAndAppendMetrics scrapes relevant metrics and appends them to
+	// the provided metric slice. If errors occur scraping some metrics,
+	// an error should be returned, but successfully scraped metrics should
+	// still be appended to the provided slice.
+	ScrapeAndAppendMetrics(ctx context.Context, metrics pdata.MetricSlice) error
```

Comment:
Maybe a different PR, I think I was wrong when I asked for append. Probably we should return a `pdata.MetricSlice` or even `pdata.Metrics` and do the append on the receiver. The reason is that `pdata.MetricSlice` is not thread safe so if for example we want to scrape in parallel it is impossible right now. I know that we have 1-2 extra allocation per scraper, but we can afford that for the benefit of being able to execute in parallel. Also the number of allocs may not even be extra because we do a Resize in every scraper. Think about it :)

Reply:
Yea I prefer that approach for both parallelizability and readability. As you point out the mumber of allocs should be similar since we were resizing the slice in each Scraper before anyway. Note in the current `ScrapeAndAppend` approach, I never bothered trying to figure out how many metrics would be produced ahead of time to minimize the number of allocations as it felt like the code would get a lot less readable (and coupled) going down that path.

I think the only reason I didn't do it this way initially is that `MoveAndAppendTo` didn't exist when I first started coding this. I figured I may as well make this change as part of this PR since we are refactoring that code anyway.

I've just updated as a separate commit. Please review the changes.